openzipkin/zipkin
# Publish to bintray

### Problem

We need to release artifacts built in travis to Bintray openzipkin/zipkin repo.  Pushing to master should trigger a snapshot release, pushing with a tag (i.e. version) would trigger a release.
### Solution
#### How it works
- user creates a tag git tag v1.2.3
- then does git push origin master --tags
- triggers the publish script to be called after travis build is successful
  - if no tag & push to master, then creates snapshot build
  - if tagged release, then pushes to bintray, updates version in gradle.properties 
- based on encrypted values in travis.yml
  - gem install travis
  - travis encrypt BINTRAY_API_KEY=<Bintray API key>--add
  - also BINTRAY_USER, GH_USER_EMAIL, GH_USER, GH_TOKEN
### Issues
- snapshot behavior not tested fully, need access from jfrog) this is done by clicking a button on the bintray package after its published
- Exception which may go away once we have snapshot access, but not sure why it appears, may be solved with access to oss.jfrog (artifactory) 

```
Execution failed for task ':zipkin-aggregate:bintrayUpload'.
> Could not create version '4.0.1-SNAPSHOT': HTTP/1.1 400 Bad Request [message:Unable to create version 4.0.1-SNAPSHOT at 'jamescway/zipkin/zipkin-aggregate']
```
### workarounds
- gradle bintrayUpload on main project is overridden to do nothing
- change git url to https instead of git://
- add git user to travis to allow pushing release commit msg from travis
- change branch to master
### To test this locally
- change all instances of openzipkin to your own org
- set env variables to your values BINTRAY_API_KEY, BINTRAY_USER, GH_USER_EMAIL, GH_USER, GH_TOKEN
- make sure you have a maven zipkin repo in bintray
- turn on github hooks in github for travis
- Test Release publishing
  - git tag 1.2.3  (any version)
  - git push origin --tags
- Test snapshot publishing
  - git push anything to master (no tags)

Fixes #521
